### DISCUSSION TITLE
Product Sorting Dropdown on All/New/Featured Listings: make available for every listing

### DISCUSSION THREAD

**Discussion by torvista** [zencart/zencart#6107] [Ideas]
![image](https://github.com/zencart/zencart/assets/4391026/34e752b4-f388-41bf-952f-a17e8aa7f186)

This is the most useful sorter: any reason it should not be available on every listing?
I lost a lot of life and hair inventing this functionality for the other listings, unaware that this one even existed.

**Comment by torvista** [comment]
It's relatively easily added to Specials, noting that the specials and featured header processing should perhaps be reviewed to make them consistent, they are different at the moment.

index listing (index_filters) is a different animal....

**Comment by torvista** [comment]
Is also relatively easily added to the index listings, if column sorting is not in use....otherwise the sort variable passed and the corresponding order_by's need to be consolidated. Not too difficult.

I find it hard to understand why this sorting functionality is not of wider interest...unless everyone uses column sorting?

**Comment by scottcwilson** [comment]
We need the King of Sort to weigh in.  @proseLA what say you?

**Comment by proseLA** [comment]
i am not the king of sort...  i'm an IT professional who enjoys coding...

i try not to get caught up in templates and presentation on the base zc.  the templates need work and could really benefit from a UI expert and much more javascript/jquery use IMO.

in looking at this sort, it looks to have been available since at least 157 (i could be wrong).  i find it hard to imagine that @torvista has not seen it before.  perhaps he has had his eyes on other things of interest to him....

i generally end up customizing sorts, as only one of my few remaining clients uses model numbers, and @torvista lives by them.  in addition, as a coder, i abhor comments like this one:

https://github.com/zencart/zencart/blob/54f714eafbdf3b3e7711f4abe58728a00bb09d83/admin/specials.php#L7

code needs to be refactored to have less code and no duplication.  on the store front, featured, new, specials and category listings could all benefit from a refactoring so that they are all consistent and the sort would be implemented once and appear on all pages.  and on the admin side, the specials and featured pages could also be refactored to make use of a single codeset.

(and i'm well aware of the refactoring progress that has been made in this arena.)

contributors could benefit from looking at what @neekfenwick has done; 38 commits, adding 2158 lines of code while removing 7206 lines of code.  that's something to be applauded IMO.

and please @scottcwilson don't ask me to do it.  i am truly still miffed at #6015 being merged.  especially when the people currently most active on this repo are all in favor of the short echo tag.  and even more so, when apparently it is good enough for the laravel code being merged into the base, but not base zc itself:
```
charmes: # grepp -l \<\?\= laravel/ 
laravel/vendor/symfony/http-kernel/Resources/welcome.html.php
laravel/vendor/symfony/error-handler/Resources/views/trace.html.php
laravel/vendor/symfony/error-handler/Resources/views/exception.html.php
laravel/vendor/symfony/error-handler/Resources/views/exception_full.html.php
laravel/vendor/symfony/error-handler/Resources/views/logs.html.php
laravel/vendor/symfony/error-handler/Resources/views/error.html.php
laravel/vendor/symfony/error-handler/Resources/views/traces.html.php
laravel/vendor/symfony/error-handler/Resources/views/traces_text.html.php
laravel/vendor/symfony/error-handler/Resources/assets/images/symfony-ghost.svg.php
```
i would apologize for the thread drift; but i will chose not to... call me out at your own peril....

**Comment by torvista** [comment]
> i find it hard to imagine that @torvista has not seen it before. perhaps he has had his eyes on other things of interest to him....

I haven't enabled the new/featured/all listings since day one, so never noticed it, much to my shock/horror.

**Reply by neekfenwick** [reply]
I ran into specials/featured etc the other week, and having worked on sort recently I thought a quick rework would be a good idea, but the code's a spaghetti mess in some places.  Styling the list of results seems inconsistent, and the way the `$list_box_contents` array is built and then used across different blocks of code gave me a real headache.  I seem to recall something was spitting out useless `<div class="">` wrapper elements around bits of the output HTML, making styling the grid of results using modern CSS like `grid` or `flexbox` a massive pain in the arse.  I gave up and avoid these listings features :)

**Reply by torvista** [reply]
Maybe there is some light in this tunnel:
https://github.com/zencart/zencart/pull/6114#issuecomment-1889804248

Long ago I had to write myself an explanation of sorting while trying to figure it out:
https://github.com/torvista/Zen_Cart-Product_Listing_Sorter/tree/master/files/docs/product_listing_sorter

It is indeed a real headache, as you have discovered and another candidate for a complete reworking to make it more easily customisable/expandable....

**Comment by neekfenwick** [comment]
> It is indeed a real headache, as you have discovered and another candidate for a complete reworking to make it more easily customisable/expandable....

As in your OP, I had no idea this more advanced sorter was available. I worry that any well meaning attempt I make to clean up the way data is used in globals between the many module and template files will run into the problem that structural changes can break an unknown number of templates in the archives that use them.  Over 12+ years I've customised, rewritten or replaced any bits of ZC that don't behave, and simply don't use the rest, so I'm not best placed to work with the core features like New, Specials, sideboxes etc.

**Reply by torvista** [reply]
Looks like you can dodge this bullet for the time being...https://github.com/zencart/zencart/pull/6154

**Reply by neekfenwick** [reply]
@torvista phew, that all looks like great work and a real relief.  I guess that work negates this whole discussion ticket?

Will look forward to it dropping into view as we pull/rebase in future.  Sounds like there is still a configurable 'number of columns' which is a shame, as I prefer to control all that via CSS and `grid`/`flex` so will have to see how useful it is to me.

**Comment by drbyte** [comment]
> On Jan 27, 2024, at 10:39 PM, neekfenwick ***@***.***> wrote:
> 
> Sounds like there is still a configurable 'number of columns' which is a shame, as I prefer to control all that via CSS and grid/flex so will have to see how useful it is to me.

That turns to "fluid" mode if the columns-per-row is set to zero, which uses grid, and is customizable without core code changes.

**Comment by neekfenwick** [comment]
> On Jan 27, 2024, at 10:39 PM, neekfenwick ***@***.***> wrote: Sounds like there is still a configurable 'number of columns' which is a shame, as I prefer to control all that via CSS and grid/flex so will have to see how useful it is to me.
> That turns to "fluid" mode if the columns-per-row is set to zero, which uses grid, and is customizable without core code changes.

Ah really? Thanks :) News to me after many years :) I googled up some docs on this at https://docs.zen-cart.com/user/template/listing_columns/ .. though it states "Note that the setting Columns Per Row only applies to Product Listing pages, not other listing pages such as All Products or New Products." and sure enough, my Configuration->Index Listing menu has e.g. "New Products Columns Per Row" but the radio buttons only allow selection of 1 through 12.  Under Product Listing, the Columns Per Row option is a textbox with no mention of 0 being a valid value or what it would do.

After investigating, I have a category page with subcats, not products.  It invokes `tpl_index_categories.php`, which uses `tpl_modules_category_row.php` which is hardcoded to use `tpl_columnar_display.php` (seems to be a copy of responsive_classic's one) which has two lines of `<div class="">` hardcoded in it.  Running on the demo database on a recent `master` branch, it results in annoying nested `div`s in the DOM, e.g.:

![image](https://github.com/zencart/zencart/assets/266415/5dab4517-6815-4b73-8394-d2829855d38f)

Buggy? Covered in another Issue?  I gave up trying to style it as it's a fairly niche section of our site we don't care much about since I redesigned the entire navigation system away from core ZC.

**Comment by scottcwilson** [comment]
> I googled up some docs on this at https://docs.zen-cart.com/user/template/listing_columns/ .. though it states "Note that the setting Columns Per Row only applies to Product Listing pages, not other listing pages such as All Products or New Products." 

This needs to be tweaked for 2.0.0.   Just updated. 

> and sure enough, my Configuration->Index Listing menu has e.g. "New Products Columns Per Row" but the radio buttons only allow selection of 1 through 12. Under Product Listing, the Columns Per Row option is a textbox with no mention of 0 being a valid value or what it would do.

Click the "?" icon at the upper right and you get help. 

Scroll down and you get the help for Columns Per Row:

```
Columns Per Row 
Key: PRODUCT_LISTING_COLUMNS_PER_ROW
Path: Configuration > Product Listing
Description: Select the number of columns of products to show per row in the product listing.
Recommended: 3
1=[rows] mode.

For the bootstrap template, use 0 (fluid columns) or 1 (rows).
```

**Reply by neekfenwick** [reply]
Fair enough.  I based my template on responsive_classic, just because it has 'responsive' in it (was very nice to move off 1.3.x table based layout!).  I've never worked with the Bootstrap one, I wouldn't know if that help text (which I hadn't read of course) would apply.

**Reply by drbyte** [reply]
@neekfenwick I'm trying to recall .... you might want to simply compare the product_listing.php module between ZC core and the BS4 template, to see if there's something additional you'd need to leverage.

**Comment by scottcwilson** [comment]
@torvista Can this be closed? The original request (add sorter to other listing pages) is done in 2.0.0.
